### DISCUSSION TITLE
[Feature request] Allow PR author to "require" all reviewers to approve

### DISCUSSION THREAD

**Discussion by nicocanali** [community/community#48162] [Pull Requests]
### Select Topic Area

Product Feedback

### Body

AFAIK, there is a setting (if you're an owner) for a minimum number of approvers, but that's not what I'm talking about. But, sometimes, PR authors (i.e. me) want to receive approval from all reviewers before merging. 

However, right now:
- 1 approval is enough to merge (in default conditions), no matter how many reviewers I add
- I have to **manually** ask reviewers  to approve it after it's already been approved by someone else (in the description of the PR for example)

Would it be something useful to explicitly mark the PR as "wants-all-reviewers" to approve or something like that?

**Comment by spalger** [comment]
I would love something like this, though I'd much prefer the ability to specify specific reviewers as required, rather than making it all or nothing. Happy to settle for all or nothing.

With large teams there are often a handful of reviewers who get added and satisfy a single "codeowner" requirement, if we get the "all or nothing" model I think it's important that a single one of the auto-added reviewers still satisfies the requirement.

My primary use-case is when I'm reviewing a PR but I want to require that design take a look before the PR auto-merges. I can mention them in my review, but there isn't any way to encode that into the PR itself right now.

A hack I've considered is adding a file which, when touched, requires a specific team's review on the PR, like:

> .github/CODEOWNERS
```
.github/require_design_review @org/ui-design
```

> .github/require_design_review
```diff
  #1234
+ #1235
```

**Reply by nicocanali** [reply]
> I'd much prefer the ability to specify specific reviewers as required

Agreed, this gives more flexibility!

**Comment by dvdsgl** [comment]
I've added a mockup:

![CleanShot 2023-03-01 at 09 39 13@2x](https://user-images.githubusercontent.com/108197/222219272-cfb6398d-f21a-4379-b250-7304c8ea9f0f.png)

**Reply by shayfrendt** [reply]
Wow @dvdsgl that looks really gorgeous. Maybe the most beautiful checkbox design I've ever seen. 🎨 

This is the only other competitor design I've seen so far:

![checkboxes](https://user-images.githubusercontent.com/32535/222220962-e6b50ca6-98d9-49a0-a23d-a25729d985ea.jpeg)

**Reply by dvdsgl** [reply]
@shayfrendt fixed that for you:

![image](https://user-images.githubusercontent.com/108197/222235168-768d427b-cbf8-4877-85cc-3fce7e96698d.png)

**Reply by nicocanali** [reply]
@dvdsgl , this is cool but would still require admin control of the repo. The optimal would be to be able to request this as a PR **author**!

But yeah, that's a start: with that setting, at least one would know that **IF** you request approval from someone, then they must approve. Other people can be CCd instead

**Comment by booiiing** [comment]
we're also looking for something like this. all-or-nothing would be fine for us. repo-admin deciding would also be fine for us.

@github team: this is a feature of BitBucket. if you implement it quickly enough, you might grab some customers. BitBucket server is coming to end of life in Feb 2024.

**Comment by booiiing** [comment]
link collection for other discussions about this topic - please feel free to reply and I will update this posting :)

- [Issue#1205](https://github.com/isaacs/github/issues/1205) on the old github project (includes solution via actions)
- [Idea#55218](https://github.com/orgs/community/discussions/55218) in this feedback hub (I reverse linked this idea there, since this is the older thread and has more participation)

**Comment by moleary-gsa** [comment]
Any chance this is being worked on?

**Comment by osievert-gpfw** [comment]
Cross posting [my duplicate feature request](https://github.com/orgs/community/discussions/55218) here to consolidate the interest and discussion in one place. I did not see this post when I made mine. Same end goals. Same suggested user experience (although I would add one more row of checkboxes to the proposed UI .. :)

Update since I made my original post: we now use a custom action to provide this functionality. It works, but has some rough edges. For example all pull requests look like they have failed until all reviewers have approved. That ends up creating a "sky is falling" scenario and now no one takes a CI failure on a PR seriously, which ruins our engineering system in a different way. The custom action has convinced me this is an important feature. But it has also convinced me that the only way to do it correctly is if it is a built-in github feature.

---

### Select Topic Area

Product Feedback

### Body

## Problem Statement

When working with github in a team setting, a lot of pull requests have a single reviewer, and as soon as that reviewer approves we merge the pull request. That makes sense. But when multiple reviewers have been selected, we sometimes find that pull requests are merged after a single reviewer has approved but another reviewer is in the middle of reviewing.

This is mostly a coordination problem, but sometimes tooling is the best way to coordinate.

Branch protection rules today allow pull requests to have a sepcified number of approvals before being merged. I currently use that feature, and have that set to "1" to ensure a review has happened. I could use this feature to ensure a second (or third, or fourth) reviewer gets a chance to complete their review, but then every pull request would require two/three/four reviews, and we have a lot of pull requests that only really need one review.

## Feature Request

Would it be possible to have another branch protection setting that is similar to this: block merging until all reviewers have approved?

## Benefit

With this new requested feature, if a pull request lists one reviewer, then only that one approval is needed in order to merge. If two reviewers were identified, both would have to approve before the pull request could be merged. This allows simple pull requests to stay simple, but ensures that more complicated pull requests get reviews from all the people who have expertise about the change (or who are impacted by the change).

## Commentary

I could see this new feature and the existing "block merging until N approvals have been given" feature interacting in a complementary way, for example:
- if no reviewers have been specified, and the minimum number of approvals is zero, then allow merging without any reviews
- if no reviewers have been specified, and the minimum number of approvals is one, then require one approval to merge (could be anyone with admin or write privilges in the repo)
- if two reviewers were specified (one of which does not have admin or write privileges), and the minimum number of approvals is three, then both specified reviewers need to approve, plus one additional approval from anyone with admin or write privileges, in order to merge

## References

This feature request is similar to, but not exactly the same, as what is discussed in  https://github.com/orgs/community/discussions/5289

**Comment by esarver** [comment]
I would also find this useful. I would like to be able to specify a list of people or teams in branch protections or rulesets that must approve before a PR can be merged. If a team is selected, it would be nice if I could specify the minimum number of people from that team that must approve.

**Comment by red2678** [comment]
+1, this would be a solid add.

**Comment by leofeyer** [comment]
We could use a "Require all reviewers to approve" checkbox, too.

**Comment by sebas-w** [comment]
+1 
![](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExd2MwanhxbHAxczJlbzlzcTZ1ZDhuenh0c3FqYnlkZTJ5am44MzdvdiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/CjNifqd0JIgV6biOjK/giphy.gif)

**Comment by MichaelGFagan** [comment]
+1, would be very happy to have this.

**Comment by darylwavy** [comment]
Either this feature or specifying mandatory vs optional reviewers would help us. Gitlab, Bitbucket and even Azure DevOps allow you to specify optional vs mandatory, so I just assumed Github had it too. When you assume...

**Comment by ryder052** [comment]
+1, this is needed

**Comment by madiej6** [comment]
+1 please add :)

**Comment by roymaor-lightspin** [comment]
+1 please add!

**Comment by rotanev** [comment]
+1, this feature should be added! If it's sensitive code that needs to be approved by all reviewers, it should be done. Such code should not be allowed to merge after one approval out of six (of three, of four, ..., for example).

**Comment by tfoskey** [comment]
+1, please add this feature. We would find it SUPER useful. :)

**Comment by NathanH-at-Hisco** [comment]
+1, please add this feature

**Comment by UkiranHisco** [comment]
+1, please add!

**Comment by sdahagam** [comment]
+1, please add this feature

**Comment by ShravanKumarASdev** [comment]
+1, Feature would be really helpful. Please add.

**Comment by alenamartinkova** [comment]
+ 1, really need it

**Comment by thimios** [comment]
+1

**Comment by shmichael** [comment]
+1

**Comment by aaronvancuren** [comment]
+1

**Comment by alexdecoder** [comment]
We have this in ADO. This is a critical feature!

**Comment by desanvicenteg** [comment]
Que Pedo con lo que se espera y quedamos ??

**Comment by melodylu** [comment]
+1

**Comment by abhimanyuPathania** [comment]
It's surprising that this is not there yet. If 2 reviewers comment on a PR. Approval from 1 commenter for their comments should not mean overall approval of the PR.
There is a setting which requires all comments to be marked `Resolved` but as of now anyone can resolve a comment. Only the one who comments should be able to resolve it.

**Comment by dalijardak** [comment]
+1

**Comment by Cwejman** [comment]
+1

**Comment by veritas-laser** [comment]
Would like to see this feature implemented.
+1

**Comment by 7138gmhampton** [comment]
Just run into a situation that would benefit from this.
+1

**Comment by melodylu** [comment]
+1

Surely there is also an open issue to have all "+1" comments convert into a direct upvote on the primary issue

**Comment by christianarg** [comment]
+100

**Comment by ktztclear987** [comment]
+1000
